Fb multi auth ui 6498

labkey-chrisj opened the pull request:
-updates to UI testing for multi-auth (particularly ldap) configuration
related PR: same branch name in ldap module

labkey-adam committed:
Attempt to fix LDAP, CAS, and SAML tests by switching to old, HTML-based configuration page

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-chrisj committed:
Item #6596 update automation for react ui in multi-auth configuration

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-chrisj committed:
Item #6596
- update to align to new UI changes
- extract common properties to authDialogBase
- add pre-cleanup to test code

labkey-tchad commented on src/org/labkey/test/components/html/ToggleButton.java:
```diff
@@ -0,0 +1,90 @@
+    final WebElement _el;
+    final WebDriver _driver;
+
+    public ToggleButton(WebElement element, WebDriver driver)
```
Make the constructor `protected` so that only the `ToggleButtonFinder` can create it.

labkey-tchad commented on src/org/labkey/test/components/html/ToggleButton.java:
```diff
@@ -0,0 +1,90 @@
+        String desiredState = enabled ? "enabled" : "disabled";
+        if (get() != enabled)
+            getComponentElement().click();
+        getWrapper().waitFor(()-> get() == enabled,
```
Pay attention to IntelliJ warnings. `waitFor` is a static method:
```suggestion
        WebDriverWrapper.waitFor(()-> get() == enabled,
```

labkey-tchad commented on src/org/labkey/test/pages/core/login/AuthDialogBase.java:
```diff
@@ -0,0 +1,97 @@
+{
+    private final LoginConfigRow _row;
+
+    protected AuthDialogBase(AuthenticationProvider<org.labkey.test.pages.ldap.LdapConfigureDialog> provider, WebDriver driver)
```
Remove reference to code in a different rep:
```suggestion
    protected AuthDialogBase(AuthenticationProvider provider, WebDriver driver)
```

labkey-tchad commented on src/org/labkey/test/pages/core/login/AuthDialogBase.java:
```diff
@@ -0,0 +1,97 @@
+    {
+//        WebElement enableToggle = Locator.tagWithClass("div", "toggle")
+//                //.withChild(Locator.tagWithClass("div", "toggle-group"))
+//                .findWhenNeeded(this).withTimeout(2000);
```
Cleanup old code

labkey-tchad commented on src/org/labkey/test/pages/core/login/AuthDialogBase.java:
```diff
@@ -0,0 +1,97 @@
+//        WebElement enableToggle = Locator.tagWithClass("div", "toggle")
+//                //.withChild(Locator.tagWithClass("div", "toggle-group"))
+//                .findWhenNeeded(this).withTimeout(2000);
+        ToggleButton enableSlider = new ToggleButton.ToggleButtonFinder(getDriver())
```
Use consistent terminology. Name this variable `enableToggle` or name the class `SliderButton`. I could, honestly, go either way.

labkey-tchad commented on src/org/labkey/test/pages/core/login/LoginConfigRow.java:
```diff
@@ -0,0 +1,115 @@
+    final WebElement _el;
+    final WebDriver _driver;
+
+    public LoginConfigRow(WebElement element, WebDriver driver)
```
Make the constructor `protected` to force usage of the `LoginConfigRowFinder`.

labkey-tchad commented on src/org/labkey/test/pages/core/login/LoginConfigRow.java:
```diff
@@ -0,0 +1,115 @@
+        final WebElement baseFieldsElement = Locator.tagWithClass("div", "domain-row-base-fields").findWhenNeeded(this);
+        final WebElement description = Locator.tagWithClass("div", "description").findElement(baseFieldsElement);
+        final WebElement details = Locator.tagWithClass("div", "details").findElement(baseFieldsElement);
+        final WebElement provider = Locator.tagWithClass("div", "provider").findElement(baseFieldsElement);
```
Any reason to not use `findWhenNeeded` for these?

labkey-chrisj replied:
nope.  it's a good idea

labkey-tchad commented on src/org/labkey/test/pages/core/login/LoginConfigurePage.java:
```diff
@@ -20,61 +22,64 @@ public LoginConfigurePage(WebDriver driver)
-        WebElement row = Locator.xpath("//tbody/tr").containing(description).findElement(elementCache().tableElement);
-        clickAndWait(Locator.tagWithClass("a", "labkey-text-link").withText("edit").findElement(row));
+        // pause, because that works?
+        sleep(1000);
```
`LoginConfigRow.clickDelete` should handle this timing.

labkey-chrisj replied:
It just happens that the timing in clickDelete has changed; deleting a configuration now puts up a prompt (which clickDelete handles, then waits for the row to become stale)
...so this sleep (a holdover from the old UI) was probably unnecessary

labkey-chrisj committed:
Item #6596
- don't reference types from optional modules/ fix build

labkey-chrisj committed:
Item #6596
- cr feedback

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-chrisj committed:
Item #6596
- updates to support CAS testing over new UI

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-chrisj committed:
Item #6596
- secondary auth configuration support

labkey-chrisj committed:
Item #6596
- extract components for sso-specifc ui

labkey-tchad commented on src/org/labkey/test/pages/core/login/AuthDialogBase.java:
```diff
@@ -0,0 +1,107 @@
+    public T clickButtonExpectingError(String buttonText)
+    {
+        Locators.dismissButton(buttonText).waitForElement(this, 2000).click();
+        return getThis();
```
`*ExpectingError` methods should wait for and return the error.

labkey-chrisj replied:
At this point, that's going to be a trick.
In the SamlConfigureDialog at least, some of the errors are at the field-level, some are text somewhere in the bottom of the dialog.  Not all of the errors are in a predictable or consistent place.

labkey-tchad replied:
Oh, that makes sense.

labkey-tchad commented on src/org/labkey/test/pages/core/login/LoginConfigurePage.java:
```diff
@@ -86,25 +159,51 @@ protected ElementCache newElementCache()
+                .findWhenNeeded(globalSettingsPanel()).withTimeout(WAIT_FOR_JAVASCRIPT);
+        WebElement autoCreateCheckBox = checkBoxLoc("Auto-create authenticated users")
+                .findWhenNeeded(globalSettingsPanel())
+                .withTimeout(WAIT_FOR_JAVASCRIPT);
```
Does the `SvgCheckbox` component not work for these ones?

labkey-chrisj replied:
What happened here... is I didn't refactor to use them here.  Good catch

labkey-tchad commented on src/org/labkey/test/tests/SecondaryAuthenticationTest.java:
```diff
@@ -1,169 +0,0 @@
-
-@Category({DailyA.class})
-@BaseWebDriverTest.ClassTimeout(minutes = 1)
-public class SecondaryAuthenticationTest extends BaseWebDriverTest
```
Why delete this test? It just needs an update for the new UI.

labkey-adam replied:
Moved to devtools (platform repo) I believe... which is where the provider lives now

labkey-chrisj replied:
it has been refactored to use the new UI, and moved into devtools, same branch name

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-chrisj committed:
Item #6596
- use svgCheckboxes in loginConfigurePage
- support for compactFileUpload component

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-chrisj committed:
Item #6596
- refactor SecurityTest test
- support for new UI for DBAuthentication
- update securityTest and testScrubber to use new UI

labkey-tchad committed:
Update test to match new modal dialog title

labkey-tchad committed:
Log accepted and dismissed alerts

labkey-chrisj committed:
Item #6596
- fix SecurityTest failure (don't click 'save and finish' expecting navigation after setting dbConfig settings)

labkey-chrisj committed:
Merge remote-tracking branch 'origin/fb_multiAuthUI_6498' into fb_multiAuthUI_6498

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-tchad commented on src/org/labkey/test/components/html/ToggleButton.java:
```diff
@@ -0,0 +1,91 @@
+    protected class ElementCache extends WebDriverComponent.ElementCache
+    {
+
+    }
```
ElementCache stuff is just taking up space. There's nothing to cache and there likely never will be.

labkey-tchad commented on src/org/labkey/test/pages/core/admin/ShowAdminPage.java:
```diff
@@ -197,7 +197,7 @@ protected ElementCache newElementCache()
         protected WebElement externalRedirectHostLink = Locator.linkWithText("External Redirect Hosts").findElement(this);
         protected WebElement auditLogLink = Locator.linkWithText("audit log").findWhenNeeded(this);
-        protected WebElement authenticationLink = Locator.linkWithText("authentication").findWhenNeeded(this);
+        protected WebElement authenticationLink = Locator.linkWithText("authentication (deprecated)").findWhenNeeded(this);
```
Remove this. Deprecated authentication page is no more.

labkey-chrisj replied:
Good catch.  Changing it to no longer use the now-absent 'authentication (deprecated)' link, to use the one that does exist instead

labkey-tchad commented on src/org/labkey/test/pages/core/login/AuthDialogBase.java:
```diff
@@ -0,0 +1,112 @@
+                Locators.dismissButton("Finish"), Locators.dismissButton("Apply")).click();
+        waitForClose(4);
+        return new LoginConfigRow.LoginConfigRowFinder(getDriver()).withDescription(description).waitFor();
+    }
```
We know whether this is a new config based on whether a `LoginConfigRow` was passed in. We could also use this knowledge to make `clickButtonExpectingError` not require button text.
```
dismiss(_row == null ? "finish" : "apply");
```

labkey-chrisj replied:
The only downside to this is that it means you don't know the button's text by just reading test code.  So, readability downside, but maintainability win

labkey-tchad commented on src/org/labkey/test/pages/core/login/DatabaseAuthConfigureDialog.java:
```diff
@@ -0,0 +1,165 @@
+            if (desiredStrength.equals(PasswordStrength.Strong))
+                elementCache().strongButton.click();
+            else
+                elementCache().weakButton.click();
```
This `if/else` is hard to read without brackets. Removing the nesting makes it more readable to me.
```
if (getPasswordStrength().equals(desiredStrength))
    return this;
else if (desiredStrength.equals(PasswordStrength.Strong))
    elementCache().strongButton.click();
else
    elementCache().weakButton.click();
```

labkey-tchad commented on src/org/labkey/test/pages/core/login/DatabaseAuthConfigureDialog.java:
```diff
@@ -0,0 +1,165 @@
+        {
+            if (oldExpiration == null)
+                oldExpiration = curExpiration;
+            setPasswordExpiration(newExpiration);
```
Remember the old settings in the individual setters or make them private. If we want to have the ability to reset we need to make it so people can't easily bypass this logic.

labkey-chrisj replied:
Okay.  I've split out the logic in that function and put the caching in the clickApply() override for this dialog

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-tchad commented on src/org/labkey/test/pages/core/login/DatabaseAuthenticationProvider.java:
```diff
@@ -0,0 +1,31 @@
+package org.labkey.test.pages.core.login;
```
This isn't a page. It should go in `org.labkey.test.params.login`, next to the base AuthenticationProvider.

labkey-chrisj committed:
Item #6596
- fix SecurityTest failure (apply changes after resetting or subsequent steps will be stuck behind the dialog mask)

labkey-tchad commented on src/org/labkey/test/TestScrubber.java:
```diff
@@ -85,16 +88,19 @@ public void cleanSiteSettings()
+            LoginConfigurePage.beginAt(this).getSecondaryConfigurations()
+                .stream().filter(a -> a.getProvider().equals("TestSecondary"))
+                    .collect(Collectors.toList())
+                    .forEach(LoginConfigRow::clickDelete);
```
Adam mentioned deleting the secondary auth configurations via API. Did that not pan out?

labkey-chrisj replied:
We did implement that via API for the particular case of deleting a secondary login config discovered during login; because that one occurs on a URL containing the config ID it was trivial

For the nonce this seemed easier to do via the UI

labkey-tchad commented on src/org/labkey/test/pages/core/login/LoginConfigurePage.java:
```diff
@@ -20,61 +23,130 @@ public LoginConfigurePage(WebDriver driver)
+    public ShowAdminPage clickSaveAndFinish()
     {
-        clickAndWait(elementCache().doneBtn());
+        shortWait().until(ExpectedConditions.elementToBeClickable(elementCache().saveAndFinishBtn()));
```
Does the save button get disabled in a way that this will notice?

labkey-chrisj replied:
not 100% sure.
Added this on the chance that it might be helpful; we may want to come up with a consistent means of knowing if a button is in a disabled state

labkey-tchad commented on src/org/labkey/test/tests/SecurityTest.java:
```diff
@@ -223,7 +224,10 @@ public void assertUrlForbidden(String url)
+        LoginConfigurePage.beginAt(this)
+                .getPrimaryConfigurationRow("Standard database authentication")
+                .clickEdit(new DatabaseAuthenticationProvider())
+                .setDbLoginConfig(DatabaseAuthConfigureDialog.PasswordStrength.Weak.Weak, null);
```
Didn't even know you could do that.
```suggestion
                .setDbLoginConfig(DatabaseAuthConfigureDialog.PasswordStrength.Weak, null);
```

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-chrisj committed:
Item #6596
- update to dismiss modal dialog rather than alert when deleting auth configuration

labkey-chrisj committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-chrisj committed:
Item #6596
- more cr feedback

labkey-chrisj committed:
Item #6596
- remove references to elementCache

labkey-chrisj committed:
Item #6596
- revert build-breaking changes

labkey-tchad committed:
Merge remote-tracking branch 'origin/develop' into fb_multiAuthUI_6498

labkey-tchad committed:
Reset DB password rules settings via API

labkey-tchad committed:
Delete authentication providers via API

labkey-tchad committed:
Fix auth dialog components

labkey-tchad committed:
Missed some configurations in auth API helper
